cpp/src/arrow/flight/server.cc

```diff
@@ -293,7 +209,36 @@ class FlightServiceImpl : public FlightService::Service {
 
   grpc::Status DoPut(ServerContext* context, grpc::ServerReader<pb::FlightData>* reader,
                      pb::PutResult* response) {
-    return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "");
+    // Get metadata
+    pb::FlightData data;
+    if (reader->Read(&data)) {
+      FlightDescriptor descriptor;
+      std::unique_ptr<ipc::Message> message;
+      std::shared_ptr<Schema> schema;
+      GRPC_RETURN_NOT_OK(internal::FromProto(data.flight_descriptor(), &descriptor));
+      std::shared_ptr<Buffer> header_buf;
+      std::shared_ptr<Buffer> body_buf;
+      GRPC_RETURN_NOT_OK(Buffer::FromString(data.data_header(), &header_buf));
+      GRPC_RETURN_NOT_OK(Buffer::FromString(data.data_body(), &body_buf));
+      GRPC_RETURN_NOT_OK(ipc::Message::Open(header_buf, body_buf, &message));
```

Comment:
Nit: this is enough boilerplate to perhaps be factored out. This might also help comparing proto-deserialization speed vs.  the zero copy optimization. A thought for a follow up JIRA perhaps

Reply:
In general, I'm a little worried by the proliferation of slightly different (de)serialization methods across Java and C++...